cloudflare.go

```diff
@@ -160,29 +160,28 @@ func (api *API) makeRequestWithAuthType(ctx context.Context, method, uri string,
 
 func (api *API) makeRequestWithAuthTypeAndHeaders(ctx context.Context, method, uri string, params interface{}, authType int, headers http.Header) ([]byte, error) {
 	// Replace nil with a JSON object if needed
-	var jsonBody []byte
+	var reqBody io.Reader
 	var err error
-	if params != nil {
-		if paramBytes, ok := params.([]byte); ok {
-			jsonBody = paramBytes
-		} else {
-			jsonBody, err = json.Marshal(params)
-			if err != nil {
-				return nil, errors.Wrap(err, "error marshalling params to JSON")
-			}
+
+	switch p := params.(type) {
+	case []byte:
+		reqBody = bytes.NewReader(p)
+	case io.Reader:
+		reqBody = p
+	case nil:
+		// empty request body
+	default:
+		jsonBody, err := json.Marshal(p)
+		if err != nil {
+			return nil, errors.Wrap(err, "error marshalling params to JSON")
 		}
-	} else {
-		jsonBody = nil
+		reqBody = bytes.NewReader(jsonBody)
 	}
 
 	var resp *http.Response
 	var respErr error
-	var reqBody io.Reader
 	var respBody []byte
 	for i := 0; i <= api.retryPolicy.MaxRetries; i++ {
-		if jsonBody != nil {
```

Comment:
If reqBody is an `io.Seeker` we may need to call [Seek](https://golang.org/src/io/io.go?s=4702:4774#L105) because the Reader could have been partially drained.

Reply:
Can we always expect a `reqBody` to be an `io.Seeker`? What if it isn't?

Is there some way we can make a copy of the `io.Reader` for each attempt? We wouldn't copy the backing value, just the pointer to the start of that value.

Reply:
It will not always be an io.Seeker. We can't copy the `io.Reader` because the whole point is to stream data instead of allocating a buffer up front. If customers want a buffer allocated up front, they can pass in a `[]byte` instead. If they use [strings.NewReader()](https://golang.org/pkg/strings/#NewReader) or [bytes.NewReader()](https://golang.org/pkg/bytes/#NewReader) they get `Seek` implemented for free. The purpose of the io.Seeker interface is to provide a way to reset the "pointer" to the start of the reader.

Reply:
can we require the parameter to be an `io.Seeker`? would like to avoid branches in logic

Reply:
We could but that prevents users from streaming data who don't want to buffer, which is a valid use case IMO. This logic is only necessary because retry logic was added inside of this rest client... I would prefer to either conditionally support io.Seeker, or to always throw an error if this library attempts to retry with an io.Reader payload. The code currently implemented gives the caller the flexibility to pass in a buffer or a stream and handles retrying gracefully when possible.

Reply:
Because large values will be supported in the future, we should fail retrying here. Remove io.Seeker support.

Reply:
@ags799 the error I would have to throw in this case must happen at runtime, even if we remove the `io.Seeker` support. Because of this, I don't think the error will be less confusing by removing `io.Seeker` support. I think we should keep this in because I think that most customers are going to pass in `bytes.NewReader()` or `strings.NewReader()` which already support `io.Seeker`... giving them retry support for free.

Reply:
Switching back to `[]byte` because this API likely won't need to support large files.